stairway-azure/src/main/java/bio/terra/stairway/azure/AzureServiceBusQueue.java

```diff
@@ -0,0 +1,240 @@
+   * @param serviceBusSenderClient the sender client
+   */
+  AzureServiceBusQueue(
+      ServiceBusReceiverClient serviceBusReceiverClient,
+      ServiceBusSenderClient serviceBusSenderClient) {
+    this.serviceBusReceiverClient = serviceBusReceiverClient;
+    this.serviceBusSenderClient = serviceBusSenderClient;
+  }
+
+  /**
+   * Construct the queue by creating a receiver and sender client using the builder parameters.
+   *
+   * @param builder the builder used to pass parameters
+   */
+  public AzureServiceBusQueue(AzureServiceBusQueue.Builder builder) {
+    ServiceBusClientBuilder serviceBusClientBuilder;
+
+    if (builder.useManagedIdentity) {
+      serviceBusClientBuilder =
+          new ServiceBusClientBuilder()
+              .fullyQualifiedNamespace(builder.namespace)
+              .credential(new DefaultAzureCredentialBuilder().build());
+    } else {
+      serviceBusClientBuilder =
+          new ServiceBusClientBuilder().connectionString(builder.connectionString);
+    }
+
+    serviceBusReceiverClient =
+        serviceBusClientBuilder
+            .receiver()
+            .topicName(builder.topicName)
+            .subscriptionName(builder.subscriptionName)
+            .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
+            .disableAutoComplete()
+            .maxAutoLockRenewDuration(builder.maxAutoLockRenewDuration)
+            .buildClient();
+
+    serviceBusSenderClient =
+        serviceBusClientBuilder.sender().topicName(builder.topicName).buildClient();
+  }
+
+  /**
+   * Reads a message from Service Bus using the receiver client. The receive operation uses
+   * peek-lock semantics. The message is acknowledged after it has been successfully processed. If
+   * the message fails to be processed, the message will be abandoned (placed back on the
+   * queue/subscription).
+   */
+  @Override
+  public void dispatchMessages(int maxMessages, QueueProcessFunction processFunction)
+      throws InterruptedException {
+
+    // Iterating over a list instead of a stream().forEach() so that the interrupted exception can
+    // be re-thrown
+    List<ServiceBusReceivedMessage> messages =
+        serviceBusReceiverClient.receiveMessages(maxMessages).stream().toList();
+
+    for (ServiceBusReceivedMessage message : messages) {
+      ProcessMessage(processFunction, message);
+    }
+  }
+
+  private void ProcessMessage(
+      QueueProcessFunction processFunction, ServiceBusReceivedMessage message)
+      throws InterruptedException {
+    try {
+      // toString uses StandardCharsets.UTF_8 by default
+      boolean processSucceeded = processFunction.apply(message.getBody().toString());
+
+      if (processSucceeded) {
+        serviceBusReceiverClient.complete(message);
+        logger.info("Completed message. ID: {}", message.getMessageId());
+        return;
+      }
+
+      logger.info("Failed to process message. ID: {}", message.getMessageId());
+      serviceBusReceiverClient.abandon(message);
```

Comment:
Does `abandon` cause the message to get reprocessed by the queue? (which is the semantics we want, I think)

Reply:
Correct, it returns the message to the queue.